consensus/polybft/slashing/double_signing_tracker.go

```diff
@@ -0,0 +1,189 @@
+package slashing
+
+import (
+	"bytes"
+
+	ibftProto "github.com/0xPolygon/go-ibft/messages/proto"
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/0xPolygon/polygon-edge/consensus/polybft/wallet"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type SenderMessagesMap map[types.Address][]*ibftProto.Message
+
+type DoubleSignEvidence struct {
+	signer   types.Address
+	round    uint64
+	messages []*ibftProto.Message
+}
+
+func newDoubleSignEvidence(signer types.Address, round uint64,
+	messages []*ibftProto.Message) *DoubleSignEvidence {
+	return &DoubleSignEvidence{signer: signer, round: round, messages: messages}
+}
+
+type MessagesMap map[uint64]map[uint64]SenderMessagesMap
+
+// getSenderMsgs returns commit messages for given height, round and sender
+func (m MessagesMap) getSenderMsgs(view *ibftProto.View, sender types.Address) []*ibftProto.Message {
+	if roundMsgs, ok := m[view.Height]; ok {
+		if senderMsgsMap, ok := roundMsgs[view.Round]; ok {
+			return senderMsgsMap[sender]
+		}
+	}
+
+	return nil
+}
+
+// registerSenderMsgs registers messages for the given height, round and sender
+func (m MessagesMap) registerSenderMsgs(view *ibftProto.View, sender types.Address, msgs []*ibftProto.Message) {
+	var senderMsgs SenderMessagesMap
+
+	if roundMsgs, ok := m[view.Height]; !ok {
+		roundMsgs = make(map[uint64]SenderMessagesMap)
+		m[view.Height] = roundMsgs
+		senderMsgs = createSenderMsgsMap(view.Round, roundMsgs)
+	} else if senderMsgs, ok = roundMsgs[view.Round]; !ok {
+		senderMsgs = createSenderMsgsMap(view.Round, roundMsgs)
+	}
+
+	senderMsgs[sender] = msgs
+}
+
+// createSenderMsgsMap initializes senders message map for the given round
+func createSenderMsgsMap(round uint64, roundMsgs map[uint64]SenderMessagesMap) SenderMessagesMap {
+	sendersMsgs := make(SenderMessagesMap)
+	roundMsgs[round] = sendersMsgs
+
+	return sendersMsgs
+}
+
+type DoubleSigningTracker interface {
+	Handle(msg *ibftProto.Message)
+	GetEvidences(height uint64) []*DoubleSignEvidence
+	PruneMsgsUntil(height uint64)
+}
+
+type DoubleSigningTrackerImpl struct {
+	preprepare  MessagesMap
+	prepare     MessagesMap
+	commit      MessagesMap
+	roundChange MessagesMap
+
+	logger hclog.Logger
+}
+
+func NewDoubleSigningTracker(logger hclog.Logger) *DoubleSigningTrackerImpl {
+	return &DoubleSigningTrackerImpl{
+		logger:      logger,
+		preprepare:  make(MessagesMap),
+		prepare:     make(MessagesMap),
+		commit:      make(MessagesMap),
+		roundChange: make(MessagesMap),
+	}
+}
+
+// Handle is implementation of IBFTMessageHandler interface, which handles IBFT consensus messages
+func (t *DoubleSigningTrackerImpl) Handle(msg *ibftProto.Message) {
+	signer, err := wallet.RecoverSignerFromIBFTMessage(msg)
+	if err != nil {
+		t.logger.Warn("failed to recover signer upon message receive",
```

Comment:
Maybe we should log this on Debug or Info level? Since the error is not that important for system functioning. Same goes for line 100.

Reply:
Fixed https://github.com/0xPolygon/polygon-edge/pull/1756/commits/9699327c17a940f08b201d0d0408cab6879d50f4